// eslint-plugin-import-zod@1.2.1/dist/index.js
//#region rolldown:runtime
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));

//#endregion
const __typescript_eslint_utils = __toESM(require("@typescript-eslint/utils"));

//#region src/rules/prefer-zod-namespace.ts
const createRule = __typescript_eslint_utils.ESLintUtils.RuleCreator((name) => `https://github.com/samchungy/eslint-plugin-import-zod/blob/main/docs/rules/${name}.md`);
const getImportSource = (importedName, originalSource) => {
	return importedName === "core" && originalSource === "zod/v4" ? `${originalSource}/${importedName}` : originalSource;
};
const isAllSameSubmodule = (specifiers, targetName) => {
	return specifiers.every((s) => s.imported.type === __typescript_eslint_utils.TSESTree.AST_NODE_TYPES.Identifier && s.imported.name === targetName);
};
var prefer_zod_namespace_default = createRule({
	name: "prefer-zod-namespace",
	meta: {
		type: "suggestion",
		docs: { description: "Enforce using namespace imports for zod" },
		fixable: "code",
		schema: [],
		messages: { preferNamespaceImport: "Import zod as a namespace (import * as z from \"zod\") instead of destructuring its exports or using default imports" }
	},
	defaultOptions: [],
	create(context) {
		return { ImportDeclaration(node) {
			if (node.source.value !== "zod" && !node.source.value.startsWith("zod/")) return;
			const namedSpecifiers = node.specifiers.filter((specifier) => specifier.type === __typescript_eslint_utils.TSESTree.AST_NODE_TYPES.ImportSpecifier);
			const defaultSpecifiers = node.specifiers.filter((specifier) => specifier.type === __typescript_eslint_utils.TSESTree.AST_NODE_TYPES.ImportDefaultSpecifier);
			if (namedSpecifiers.length === 0 && defaultSpecifiers.length === 0) return;
			const zodSpecifiers = namedSpecifiers.filter((specifier) => specifier.imported.type === __typescript_eslint_utils.TSESTree.AST_NODE_TYPES.Identifier && (specifier.imported.name === "z" || specifier.imported.name === "core"));
			const zodDefaultSpecifiers = defaultSpecifiers;
			if (zodSpecifiers.length === 0 && zodDefaultSpecifiers.length === 0) return;
			for (const zodDefaultSpecifier of zodDefaultSpecifiers) {
				const localName = zodDefaultSpecifier.local.name;
				const importSource = node.source.value;
				context.report({
					node: zodDefaultSpecifier,
					messageId: "preferNamespaceImport",
					fix(fixer) {
						if (node.specifiers.length === 1) {
							const isTypeOnlyImport = node.importKind === "type";
							const typePrefix = isTypeOnlyImport ? "type " : "";
							return fixer.replaceText(node, `import ${typePrefix}* as ${localName} from '${importSource}';`);
						} else {
							const otherSpecifiers = node.specifiers.filter((s) => s !== zodDefaultSpecifier);
							const isTypeOnlyImport = node.importKind === "type";
							const typePrefix = isTypeOnlyImport ? "type " : "";
							const namespaceImport = `import ${typePrefix}* as ${localName} from '${importSource}';\n`;
							const otherImportParts = [];
							const otherDefaultSpecifiers = otherSpecifiers.filter((s) => s.type === __typescript_eslint_utils.TSESTree.AST_NODE_TYPES.ImportDefaultSpecifier);
							const otherNamedSpecifiers = otherSpecifiers.filter((s) => s.type === __typescript_eslint_utils.TSESTree.AST_NODE_TYPES.ImportSpecifier);
							if (otherDefaultSpecifiers.length > 0) {
								const defaultName = otherDefaultSpecifiers[0].local.name;
								otherImportParts.push(defaultName);
							}
							if (otherNamedSpecifiers.length > 0) {
								const namedPart = `{ ${otherNamedSpecifiers.map((s) => {
									const specifierLocalName = s.local.name;
									const specifierImportedName = s.imported.type === __typescript_eslint_utils.TSESTree.AST_NODE_TYPES.Identifier ? s.imported.name : "";
									const typeModifier = !isTypeOnlyImport && s.importKind === "type" ? "type " : "";
									return specifierLocalName === specifierImportedName ? `${typeModifier}${specifierImportedName}` : `${typeModifier}${specifierImportedName} as ${specifierLocalName}`;
								}).join(", ")} }`;
								otherImportParts.push(namedPart);
							}
							const otherImport = `import ${isTypeOnlyImport ? "type " : ""}${otherImportParts.join(", ")} from '${importSource}';`;
							return fixer.replaceText(node, `${namespaceImport}${otherImport}`);
						}
					}
				});
			}
			for (const zodSpecifier of zodSpecifiers) {
				const localName = zodSpecifier.local.name;
				const importedName = zodSpecifier.imported.name;
				const importSource = getImportSource(importedName, node.source.value);
				const isSubmoduleImport = importSource !== node.source.value;
				context.report({
					node: zodSpecifier,
					messageId: "preferNamespaceImport",
					fix(fixer) {
						if (namedSpecifiers.length === 1 || isSubmoduleImport && isAllSameSubmodule(namedSpecifiers, importedName)) {
							const isTypeOnlyImport = node.importKind === "type";
							const typePrefix = isTypeOnlyImport ? "type " : "";
							return fixer.replaceText(node, `import ${typePrefix}* as ${localName} from '${importSource}';`);
						} else {
							const otherSpecifiers = namedSpecifiers.filter((s) => s !== zodSpecifier);
							const isTypeOnlyImport = node.importKind === "type";
							const typePrefix = isTypeOnlyImport ? "type " : "";
							const namespaceImport = `import ${typePrefix}* as ${localName} from '${importSource}';\n`;
							const originalSource = node.source.value;
							const otherImport = `import ${isTypeOnlyImport ? "type " : ""}{ ${otherSpecifiers.map((s) => {
								const specifierLocalName = s.local.name;
								const specifierImportedName = s.imported.type === __typescript_eslint_utils.TSESTree.AST_NODE_TYPES.Identifier ? s.imported.name : "";
								const typeModifier = !isTypeOnlyImport && s.importKind === "type" ? "type " : "";
								return specifierLocalName === specifierImportedName ? `${typeModifier}${specifierImportedName}` : `${typeModifier}${specifierImportedName} as ${specifierLocalName}`;
							}).join(", ")} } from '${originalSource}';`;
							return fixer.replaceText(node, `${namespaceImport}${otherImport}`);
						}
					}
				});
			}
		} };
	}
});

//#endregion
//#region src/index.ts
const importZod = {
	meta: {
		name: "import-zod",
		version: "1.0.0"
	},
	rules: { "prefer-zod-namespace": prefer_zod_namespace_default },
	configs: {}
};
Object.assign(importZod.configs, { recommended: [{
	plugins: { "import-zod": importZod },
	rules: { "import-zod/prefer-zod-namespace": "error" }
}] });
module.exports = importZod;

//#endregion
